### DISCUSSION TITLE
Add support for exporting to Android from the Android editor

### DISCUSSION THREAD

**Discussion by anonimo82** [godotengine/godot-proposals#4324] [Platforms]
Since Godot editor now runs on Android (https://downloads.tuxfamily.org/godotengine/3.5/beta3/), and Android deployment needs a valid SDK, maybe you can consider supporting AndroidIDE (https://github.com/itsaky/AndroidIDE - https://androidide.com/) which provides a minimal Android SDK, to compile Android projects, inside Android!!! =)

Also see here (https://github.com/itsaky/AndroidIDE/issues/148)

Cheers =)

**Comment by Calinou** [comment]
In principle, if you use the [custom build system](https://docs.godotengine.org/en/stable/tutorials/export/android_custom_build.html), it should be possible to import the exported project files into AndroidIDE.

However, quoting the [Android IDE README](https://github.com/itsaky/AndroidIDE#limitations):

> But, you cannot use it to install some tools (like NDK) because those tools are not built for Android.

Building a Godot APK traditionally requires the NDK to be available. I'm not sure if this step can be bypassed if you've already got access to precompiled libraries (which custom builds already include).

**Reply by anonimo82** [reply]
Tried to export the custom build, the editor has crashed
Edit: Now I can't even try to export, I don't have the option anymore

**Reply by anonimo82** [reply]
Ok, so, I got project export almost working, but not fully.
I unzipped Android SDK from AndroidIDE and used its path in Godot, I generated a Debug KeyStore, but exporting the project ends with a build error, even if I'm trying to just export code (am I?)

Please, help

Edit: Attaching screenshot https://pasteboard.co/5crfQWSUok8i.jpg

Edit 2: Ok, I made it to export source code and Android project. Now, I'm trying to build with AIDE (https://play.google.com/store/apps/details?id=com.aide.ui), because it supports NDK.
Here is what I get: https://pasteboard.co/NQtLYMKCUwlI.jpg

**Comment by anonimo82** [comment]
Uhm, no, we really need some kind of integration with the IDE, because it's not even possible to run Gradle from the editor.
Maybe not strictly the one I linked here, but we at least need to be able to run Gradle to set up the project.
An alternative could be to fork some Android SDK port and put inside Godot

**Reply by Calinou** [reply]
Android does not allow running arbitrary binaries with `OS.execute()`, so I don't see how this could be done.

> An alternative could be to fork some Android SDK port and put inside Godot

The Android SDK is way too large to consider doing that. To make things worse, the full Android SDK distribution is proprietary – only its individual components are open source. We'd have to put the components together one by one, which is a lot of work. To my knowledge, nobody in the Android community managed to do this for recent SDK versions.

**Comment by anonimo82** [comment]
Dunno what to say, AndroidIDE has a subset of the SDK, it's also gonna support Kotlin, but not (yet) NDK.
The limit is you have to name your Android project "app", at the moment.
Otherwise, you could target AIDE (https://www.android-ide.com/ - note this URL is similar to the other IDE's one) which supports NDK, but you'll need to already have dependencies downloaded, because it lacks a good Gradle support, and you'll have no Kotlin. Other IDEs/SDKs are just worthless IMHO.

Edit: Note my idea is actually to export a suitable project, then build it with some Android IDE

**Comment by anonimo82** [comment]
Any new toughts on this?

**Reply by Calinou** [reply]
No, there isn't anything we can do about this to be fair. AIDE doesn't appear to be updated often, which doesn't help with the Android exporting situation (as Google changes the requirements very often).

**Reply by anonimo82** [reply]
Yes, I see.
What about Android IDE?
It just lacks NDK support for now, and is actively developed. 
https://github.com/itsaky/AndroidIDE
https://androidide.com/
It's also on its way to support Kotlin.

**Reply by anonimo82** [reply]
I'm trying to build a Godot project with Android IDE, but I'd need the main project to be named "app".
What should I change in the project's code?
Also note it came to my knowledge there's a non official way to achieve NDK in Android IDE, so may be this would definitely be the IDE to go with.
NDK link: https://github.com/Lzhiyong/termux-ndk

**Comment by anonimo82** [comment]
Hey there, Android IDE's author helped me a lot and I finally made it to build and sign an APK from a Godot generated project, then built with Android IDE.
On screen controls doesn't work, and I can't hear audio, but controls neither works inside Godot, so I'm not very concerned, by the way, audio should work and it doesn't.
Android IDE's author is on the way to fix APK signing problem, I used a .keystore I generated from my PC, ATM.
Android's editor port needs some tweaking, because on executing a project, it still acts as if it were a PC, while it should act like a smartphone or tablet.

For the records, I built the 3D platformer demo from the template in the content library.

**Reply by anonimo82** [reply]
P.S.: Look at the other (linked at top) issue to see how I successfully built an APK.
Effort should be put in streamlining the workflow as much as possible on both sides, Godot and Android IDE

**Comment by Calinou** [comment]
This is supported since 4.4, but only for APK exporting, not Gradle or AAB. See the description of https://github.com/godotengine/godot/pull/93526 for details.
